File: crates/headgate/src/task_data.rs
//! In-process typed data for workers and job attempts.
//!
//! This is deliberately not part of [`headgate_core::Envelope`]. Worker data is shared
//! by every attempt run by one worker; job data is a fresh map for each attempt and is
//! shared only by clones of that attempt's [`crate::JobCtx`]. Values are keyed by Rust's
//! [`TypeId`], not by strings, and returned behind [`Arc`] so a handler never holds a
//! lock across an `.await`.

use std::any::{Any, TypeId};
use std::collections::HashMap;
use std::sync::{Arc, RwLock};

type Value = Arc<dyn Any + Send + Sync>;

/// A concurrency-safe heterogeneous type map.
///
/// Cloning `Extensions` clones the handle to the same map. A worker owns one shared
/// instance; the runtime creates a different empty instance for every job attempt.
/// Neither the container nor its values implement headgate's wire types, so nothing in
/// this map is serialized into a job envelope.
#[derive(Clone, Default)]
pub struct Extensions {
    values: Arc<RwLock<HashMap<TypeId, Value>>>,
}

impl Extensions {
    pub fn new() -> Self {
        Self::default()
    }

    /// Insert one value under its concrete type, returning the previous value of that
    /// exact type. A value of another type can neither collide nor be downcast by the
    /// caller.
    pub fn insert<T>(&self, value: T) -> Option<Arc<T>>
    where
        T: Send + Sync + 'static,
    {
        self.values
            .write()
            .unwrap()
            .insert(TypeId::of::<T>(), Arc::new(value))
            .and_then(|old| Arc::downcast::<T>(old).ok())
    }

    /// Get a shared handle to the value stored under exactly `T`.
    pub fn get<T>(&self) -> Option<Arc<T>>
    where
        T: Send + Sync + 'static,
    {
        self.values
            .read()
            .unwrap()
            .get(&TypeId::of::<T>())
            .cloned()
            .and_then(|value| Arc::downcast::<T>(value).ok())
    }

    pub fn remove<T>(&self) -> Option<Arc<T>>
    where
        T: Send + Sync + 'static,
    {
        self.values
            .write()
            .unwrap()
            .remove(&TypeId::of::<T>())
            .and_then(|value| Arc::downcast::<T>(value).ok())
    }

    pub fn contains<T>(&self) -> bool
    where
        T: Send + Sync + 'static,
    {
        self.values.read().unwrap().contains_key(&TypeId::of::<T>())
    }

    pub fn len(&self) -> usize {
        self.values.read().unwrap().len()
    }

    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }
}

File: go/task_data.go
package headgate

// Type-safe, in-process data for workers and job attempts.
//
// The map is keyed by reflect.Type rather than a string. A typed box preserves
// even a typed nil value without an unsafe cast. Generic methods and the original
// package functions share the same storage and synchronization.

import (
	"context"
	"errors"
	"reflect"
	"sync"
)

// ErrTaskDataUnavailable means a task-data operation was attempted outside a handler
// context. Worker data is configured explicitly through Config.Extensions instead.
var ErrTaskDataUnavailable = errors.New("headgate: task data is only available inside a handler")

// Extensions is a concurrency-safe heterogeneous type map. It is process-local and is
// not a field of Envelope: no extension can be serialized into a queued job by the
// runtime. Pass one instance in Config.Extensions for worker-shared dependencies; the
// runner creates a different empty instance for every job attempt.
type Extensions struct {
	mu     sync.RWMutex
	values map[reflect.Type]any
}

func NewExtensions() *Extensions { return &Extensions{} }

// Set stores value under exactly T and returns its previous value, if present.
// Like SetExtension, it panics when extensions is nil.
func (extensions *Extensions) Set[T any](value T) (previous T, replaced bool) {
	return SetExtension[T](extensions, value)
}

// Get returns the value stored under exactly T. A nil receiver or missing type
// returns the zero value and false.
func (extensions *Extensions) Get[T any]() (value T, ok bool) {
	return Extension[T](extensions)
}

// Remove deletes and returns the value stored under exactly T. A nil receiver or
// missing type returns the zero value and false.
func (extensions *Extensions) Remove[T any]() (value T, ok bool) {
	return RemoveExtension[T](extensions)
}

type extensionBox[T any] struct{ value T }

func extensionType[T any]() reflect.Type { return reflect.TypeOf((*T)(nil)).Elem() }

// SetExtension stores value under exactly T and returns the prior value of T, if any.
func SetExtension[T any](extensions *Extensions, value T) (previous T, replaced bool) {
	if extensions == nil {
		panic("headgate: SetExtension called with nil Extensions")
	}
	extensions.mu.Lock()
	defer extensions.mu.Unlock()
	if extensions.values == nil {
		extensions.values = make(map[reflect.Type]any)
	}
	key := extensionType[T]()
	old, replaced := extensions.values[key]
	extensions.values[key] = extensionBox[T]{value: value}
	if !replaced {
		return previous, false
	}
	return old.(extensionBox[T]).value, true
}

// Extension returns the value stored under exactly T. Asking for another type is a
// miss; callers never receive an untyped value to cast.
func Extension[T any](extensions *Extensions) (value T, ok bool) {
	if extensions == nil {
		return value, false
	}
	extensions.mu.RLock()
	defer extensions.mu.RUnlock()
	boxed, ok := extensions.values[extensionType[T]()]
	if !ok {
		return value, false
	}
	return boxed.(extensionBox[T]).value, true
}

func RemoveExtension[T any](extensions *Extensions) (value T, ok bool) {
	if extensions == nil {
		return value, false
	}
	extensions.mu.Lock()
	defer extensions.mu.Unlock()
	key := extensionType[T]()
	boxed, ok := extensions.values[key]
	if !ok {
		return value, false
	}
	delete(extensions.values, key)
	return boxed.(extensionBox[T]).value, true
}

func (extensions *Extensions) Len() int {
	if extensions == nil {
		return 0
	}
	extensions.mu.RLock()
	defer extensions.mu.RUnlock()
	return len(extensions.values)
}

type taskDataScope struct {
	worker *Extensions
	job    *Extensions
}

type taskDataContextKey struct{}

func withTaskData(ctx context.Context, worker *Extensions) context.Context {
	if worker == nil {
		worker = NewExtensions()
	}
	return context.WithValue(ctx, taskDataContextKey{}, taskDataScope{
		worker: worker,
		job:    NewExtensions(),
	})
}

func taskDataFrom(ctx context.Context) (taskDataScope, bool) {
	data, ok := ctx.Value(taskDataContextKey{}).(taskDataScope)
	return data, ok
}

// SetJobData inserts scratch data into this attempt only. Concurrent jobs always have
// different job maps; derived contexts and goroutines for this job share the same map.
func SetJobData[T any](ctx context.Context, value T) error {
	data, ok := taskDataFrom(ctx)
	if !ok {
		return ErrTaskDataUnavailable
	}
	SetExtension(data.job, value)
	return nil
}

func JobData[T any](ctx context.Context) (value T, ok bool) {
	data, ok := taskDataFrom(ctx)
	if !ok {
		return value, false
	}
	return Extension[T](data.job)
}

func WorkerData[T any](ctx context.Context) (value T, ok bool) {
	data, ok := taskDataFrom(ctx)
	if !ok {
		return value, false
	}
	return Extension[T](data.worker)
}

// Data resolves T from this attempt first, then from the worker's shared defaults.
// Job-local shadowing lets middleware specialize one dependency without mutating the
// value seen by concurrently-running siblings.
func Data[T any](ctx context.Context) (value T, ok bool) {
	if value, ok = JobData[T](ctx); ok {
		return value, true
	}
	return WorkerData[T](ctx)
}
